File: src/pipeline/segmenter.ts
/**
 * Step 1B — Segmenter.
 * fixed_window: blocks of 120-240s (config)
 * topic_shift: semantic segmentation with max ~12 segments/video (naive sentence-split for now)
 */

import type { CleanSegment } from "./transcript-cleaner.js";

export interface Segment {
  segment_id: string;
  start_sec: number;
  end_sec: number;
  text: string;
  token_count: number;
}

export type SegmenterMode = "fixed_window" | "topic_shift";

export interface SegmenterConfig {
  mode: SegmenterMode;
  /** fixed_window: target duration in seconds */
  window_seconds?: number;
  /** fixed_window: min overlap to avoid cutting mid-sentence */
  min_window_seconds?: number;
  /** topic_shift: max segments per video */
  max_segments?: number;
}

const DEFAULT_FIXED: SegmenterConfig = {
  mode: "fixed_window",
  window_seconds: 180,
  min_window_seconds: 120,
};

const DEFAULT_TOPIC: SegmenterConfig = {
  mode: "topic_shift",
  max_segments: 12,
};

function countTokens(text: string): number {
  return text.split(/\s+/).filter(Boolean).length;
}

function segmentFixedWindow(
  items: CleanSegment[],
  windowSec: number,
  minWindowSec: number
): Segment[] {
  const segments: Segment[] = [];
  let segIdx = 0;
  let currentStart = 0;
  let currentEnd = 0;
  let currentText: string[] = [];
  let currentTokens = 0;

  for (const item of items) {
    const itemEnd = item.end_sec;
    const itemTokens = countTokens(item.text);

    const wouldExceed = (currentEnd > 0 && itemEnd - currentStart > windowSec) ||
      (currentEnd > 0 && itemEnd - currentStart >= minWindowSec && currentTokens + itemTokens > 100);

    if (wouldExceed && currentText.length > 0) {
      segments.push({
        segment_id: `s${segIdx}`,
        start_sec: currentStart,
        end_sec: currentEnd,
        text: currentText.join(" ").trim(),
        token_count: currentTokens,
      });
      segIdx++;
      currentStart = item.start_sec;
      currentEnd = item.end_sec;
      currentText = [item.text];
      currentTokens = itemTokens;
    } else {
      if (currentText.length === 0) currentStart = item.start_sec;
      currentEnd = item.end_sec;
      currentText.push(item.text);
      currentTokens += itemTokens;
    }
  }

  if (currentText.length > 0) {
    segments.push({
      segment_id: `s${segIdx}`,
      start_sec: currentStart,
      end_sec: currentEnd,
      text: currentText.join(" ").trim(),
      token_count: currentTokens,
    });
  }

  return segments;
}

/** topic_shift: naive implementation — split by chunk size to hit max_segments */
function segmentTopicShift(
  items: CleanSegment[],
  maxSegments: number
): Segment[] {
  const totalText = items.map((x) => x.text).join(" ");
  const totalTokens = countTokens(totalText);
  const targetPerSeg = Math.max(50, Math.ceil(totalTokens / maxSegments));

  const segments: Segment[] = [];
  let segIdx = 0;
  let accText: string[] = [];
  let accTokens = 0;
  let segStart = 0;
  let segEnd = 0;

  for (const item of items) {
    const tokens = countTokens(item.text);
    const wouldExceed = accTokens + tokens > targetPerSeg && accText.length > 0;

    if (wouldExceed && segIdx < maxSegments - 1) {
      segments.push({
        segment_id: `s${segIdx}`,
        start_sec: segStart,
        end_sec: segEnd,
        text: accText.join(" ").trim(),
        token_count: accTokens,
      });
      segIdx++;
      accText = [item.text];
      accTokens = tokens;
      segStart = item.start_sec;
      segEnd = item.end_sec;
    } else {
      if (accText.length === 0) segStart = item.start_sec;
      accText.push(item.text);
      accTokens += tokens;
      segEnd = item.end_sec;
    }
  }

  if (accText.length > 0) {
    segments.push({
      segment_id: `s${segIdx}`,
      start_sec: segStart,
      end_sec: segEnd,
      text: accText.join(" ").trim(),
      token_count: accTokens,
    });
  }

  return segments;
}

export function segment(
  cleanSegments: CleanSegment[],
  config: Partial<SegmenterConfig> = {}
): Segment[] {
  const cfg = config.mode === "topic_shift"
    ? { ...DEFAULT_TOPIC, ...config }
    : { ...DEFAULT_FIXED, ...config };

  if (cleanSegments.length === 0) return [];

  if (cfg.mode === "fixed_window") {
    return segmentFixedWindow(
      cleanSegments,
      cfg.window_seconds ?? 180,
      cfg.min_window_seconds ?? 120
    );
  }

  return segmentTopicShift(
    cleanSegments,
    cfg.max_segments ?? 12
  );
}

File: src/media_advisor/pipeline/segmenter.py
"""Step 1B — Segmenter.

fixed_window: blocks of ~180s.
topic_shift: naive token-count split targeting max_segments per video.
Porting of src/pipeline/segmenter.ts.
"""

from dataclasses import dataclass
from typing import Literal

from media_advisor.pipeline.cleaner import CleanSegment

SegmenterMode = Literal["fixed_window", "topic_shift"]


@dataclass
class Segment:
    segment_id: str
    start_sec: float
    end_sec: float
    text: str
    token_count: int


def _count_tokens(text: str) -> int:
    return len([w for w in text.split() if w])


def _segment_fixed_window(
    items: list[CleanSegment],
    window_sec: float,
    min_window_sec: float,
) -> list[Segment]:
    segments: list[Segment] = []
    seg_idx = 0
    current_start = 0.0
    current_end = 0.0
    current_text: list[str] = []
    current_tokens = 0

    for item in items:
        item_end = item.end_sec
        item_tokens = _count_tokens(item.text)

        would_exceed = (current_end > 0 and item_end - current_start > window_sec) or (
            current_end > 0
            and item_end - current_start >= min_window_sec
            and current_tokens + item_tokens > 100
        )

        if would_exceed and current_text:
            segments.append(
                Segment(
                    segment_id=f"s{seg_idx}",
                    start_sec=current_start,
                    end_sec=current_end,
                    text=" ".join(current_text).strip(),
                    token_count=current_tokens,
                )
            )
            seg_idx += 1
            current_start = item.start_sec
            current_end = item_end
            current_text = [item.text]
            current_tokens = item_tokens
        else:
            if not current_text:
                current_start = item.start_sec
            current_end = item_end
            current_text.append(item.text)
            current_tokens += item_tokens

    if current_text:
        segments.append(
            Segment(
                segment_id=f"s{seg_idx}",
                start_sec=current_start,
                end_sec=current_end,
                text=" ".join(current_text).strip(),
                token_count=current_tokens,
            )
        )

    return segments


def _segment_topic_shift(items: list[CleanSegment], max_segments: int) -> list[Segment]:
    total_text = " ".join(x.text for x in items)
    total_tokens = _count_tokens(total_text)
    target_per_seg = max(50, -(-total_tokens // max_segments))  # ceiling div

    segments: list[Segment] = []
    seg_idx = 0
    acc_text: list[str] = []
    acc_tokens = 0
    seg_start = 0.0
    seg_end = 0.0

    for item in items:
        tokens = _count_tokens(item.text)
        would_exceed = acc_tokens + tokens > target_per_seg and bool(acc_text)

        if would_exceed and seg_idx < max_segments - 1:
            segments.append(
                Segment(
                    segment_id=f"s{seg_idx}",
                    start_sec=seg_start,
                    end_sec=seg_end,
                    text=" ".join(acc_text).strip(),
                    token_count=acc_tokens,
                )
            )
            seg_idx += 1
            acc_text = [item.text]
            acc_tokens = tokens
            seg_start = item.start_sec
            seg_end = item.end_sec
        else:
            if not acc_text:
                seg_start = item.start_sec
            acc_text.append(item.text)
            acc_tokens += tokens
            seg_end = item.end_sec

    if acc_text:
        segments.append(
            Segment(
                segment_id=f"s{seg_idx}",
                start_sec=seg_start,
                end_sec=seg_end,
                text=" ".join(acc_text).strip(),
                token_count=acc_tokens,
            )
        )

    return segments


def segment(
    clean_segments: list[CleanSegment],
    mode: SegmenterMode = "fixed_window",
    window_seconds: float = 180.0,
    min_window_seconds: float = 120.0,
    max_segments: int = 12,
) -> list[Segment]:
    if not clean_segments:
        return []
    if mode == "topic_shift":
        return _segment_topic_shift(clean_segments, max_segments)
    return _segment_fixed_window(clean_segments, window_seconds, min_window_seconds)
